discovery-provider/src/challenges/challenge_event_bus.py

```diff
@@ -0,0 +1,71 @@
+import json
+import enum
+import logging
+
+from collections import defaultdict
+
+logger = logging.getLogger(__name__)
+REDIS_QUEUE_PREFIX = 'challenges-event-queue'
+
+class ChallengeEvent(str, enum.Enum):
+    profile_update = 'profile_update'
+    repost = 'repost'
+    follow = 'follow'
+    favorite = 'favorite'
+
+class ChallengeEventBus:
+    """`ChallengeEventBus` supports:
+        - dispatching challenge events to a Redis queue
+        - registering challenge managers to listen to the events.
+        - consuming items from the Redis queue
+    """
+    def __init__(self, redis):
+        self._listeners = defaultdict(lambda: [])
+        self._redis = redis
+
+    def register_listener(self, event, listener):
+        """Registers a listener (`ChallengeEventManager`) to listen for a particular event type."""
+        self._listeners[event].append(listener)
+
+    def dispatch(self, session, event, block_number, user_id):
+        """Dispatches an event + block_number + user_id to Redis queue"""
+        event_json = self._event_to_json(event, block_number, user_id)
+        logger.debug(f"ChallengeEventBus: dispatch {event_json}")
+        self._redis.rpush(REDIS_QUEUE_PREFIX, event_json)
+
+    def process_events(self, session, max_events=1000):
+        """Dequeues `max_events` from Redis queue and processes them, forwarding to listening ChallengeManagers"""
+        # get the first max_events elements
+        events_json = self._redis.lrange(REDIS_QUEUE_PREFIX, 0, max_events)
+        logger.debug(f"ChallengeEventBus: dequeued {len(events_json)} events")
+        # trim the first from the front of the list
+        self._redis.ltrim(REDIS_QUEUE_PREFIX, len(events_json), -1)
+        events_dicts = list(map(self._json_to_event, events_json))
+
+        # Consolidate event types for processing
+        # map of {"event_type": [{ user_id: number, block_number: number }]}}
+        event_user_dict = defaultdict(lambda: [])
+        for event_dict in events_dicts:
+            event_type = event_dict["event"]
+            event_user_dict[event_type].append({
+                "user_id": event_dict["user_id"],
+                "block_number": event_dict["block_number"]
+            })
+
+        for (event_type, event_dicts) in event_user_dict.items():
+            listeners = self._listeners[event_type]
+            for listener in listeners:
+                listener.process(session, event_type, event_dicts)
```

Comment:
they don't strictly, but would have to be a little bit careful with how updates work on the UserChallenges now (would have to be an atomic increment in SQL). I think i'll leave this optimization for a later day, esp since this will run in a celery job anyways

Reply:
agree^